xof.md

```diff
@@ -0,0 +1,206 @@
+This could just be a new gettable flag (having a separate provider function should not be necessary).
+
+#### Proposal 1
+
+Change EVP_DigestFinalXOF(ctx, out, outlen) to handle multiple calls.
+Possibly have EVP_DigestSqueeze() just as an alias method?
+Changing the code at this level should be a simple matter of removing the flag check.
+
+##### Pros
+
+  - New API is not required
+
+##### Cons
+
+  - Final seems like a strange name to call multiple times.
+
+#### Proposal 2
+
+Keep EVP_DigestFinalXOF() as a one shot function and create a new API to handle the
+multi squeeze case e.g.
+```
+EVP_DigestSqueeze(ctx, out, outlen).
+```
+
+##### Pros
+
+  - Seems like a better name.
+
+##### Cons
+
+  - Adds an extra API.
+  - The interaction between the 2 API's needs to be clearly documented.
+     - A call to EVP_DigestSqueeze() after EVP_DigestFinalXOF() would fail.
+     - A call to EVP_DigestFinalXOF() after the EVP_DigestSqueeze() would fail? Is this confusing.
+
+API Discussion of other XOF API'S
+---------------------------------
+
+### Init
+
+The digest can be initialized as normal using:
+```
+md = EVP_MD_fetch(libctx, "SHAKE256", propq);
+ctx = EVP_MD_CTX_new();
+EVP_DigestInit_ex2(ctx, md, NULL);
+```
+
+### Absorb
+
+Absorb can be done by multiple calls to:
+```
+EVP_DigestUpdate(ctx, in, inlen);
+```
+
+#### Proposal:
+
+Do we want to have an Alias function?
+```
+EVP_DigestAbsorb(ctx, in, inlen);
+```
+
+### Finalize
+
+The finalize is just done as part of the squeeze operation.
+
+### Reset
+
+A reset can be done by calling:
+```
+EVP_DigestInit_ex2(ctx, NULL, NULL);
+```
+
+Low Level squeeze changes
+--------------------------
+
+### Description
+
+The existing one shot squeeze method is:
+```
+SHA3_squeeze(uint64_t A[5][5], unsigned char *out, size_t outlen, size_t r)
+```
+It contains an opaque object for storing the state B<A>,
+which can be used B<r> times to output squeezed values to B<out>,
+before it needs to update the state B<A> by internally calling KeccakF1600().
+Unless you are using a multiple of B<r> as the B<outlen>, the function no way
+of knowing where to start from if another call to SHA_squeeze() was attempted.
+The method also avoids doing a final call to KeccakF1600() currently since it was
+assumed that it was not required for a one shot operation.
+
+### Solution 1
+
+Modify the SHA3_squeeze code to accept a input/output parameter to track the position
+within the state B<A>.
+See https://github.com/openssl/openssl/pull/13470.
+
+#### Pros
+
+  - Change in C code is minimal. it just needs to pass this additional parameter.
+  - There are no additional memory copies of buffered results.
+
+#### Cons
+
+  - The logic in the c reference has many if clauses.
+  - It needs to be written in assembler, the logic would also be different in different assembler routines
+    due to the internal format of the state A being different.
```

Comment:
Maybe you should add something about the same change as done in the C version needs to be done in all the assembler versions.